src/main/java/com/greenfoxacademy/chamaaragonite/kingdom/KingdomServiceImpl.java

```diff
@@ -72,5 +91,135 @@ public Kingdom updateKingdom(Kingdom kingdom) {
         this.resourceService.updateResources(kingdom)
     );
   }
-  // endregion Methods
+  // endregion Kingdom
+
+
+  // region Buildings
+  @Override
+  public Kingdom createBuilding(String buildingType, String userName)
+      throws IncorrectBuildingTypeException {
+    Kingdom kingdom = this.updateKingdom(this.kingdomRepository.findByAppUserUsername(userName));
+    Integer cost = this.purchaseService.getNewBuildingPrice(buildingType);
+
+    if (cost == null) {
+      throw new IncorrectBuildingTypeException("Building type doesn't exist!");
+    }
+
+    Resource gold = kingdom.getGold();
+    if (gold.getAmount() >= cost) {
+      Building building = this.buildingService.create(buildingType);
+      kingdom.addBuilding(building);
+      gold.setAmount(
+          gold.getAmount() - cost
+      );
+      return this.updateKingdom(kingdom);
```

Comment:
```suggestion
      this.updateKingdom(kingdom);
      return building;
```
The endpoints need to respond with the created/upgraded objets's details, so wouldn't returing those be more efficient, than tryin to somehow get them from the kingdom in the controller? Only 4 small changes here, and makes the controllers simpler. 🤔💤